config.org
* configure =use-package= and path
** Do the stuff
=use-package= is 'require on [[https://duckduckgo.com/?q=usepackage+emacs&t=brave&ia=software][steroids]]. It downloads things for me. Yay.
=use-package-ensure= is equivalent to =:ensure t= on every =use-package= use.
It means install package if not already installed.
#+begin_src emacs-lisp
  (require 'use-package-ensure)
  (setq use-package-always-ensure t)
#+end_src

[[https://github.com/emacscollective/auto-compile][Auto-compile]] packages, and use the newest version available.
#+begin_src emacs-lisp
	(use-package auto-compile
		:config	(auto-compile-on-save-mode) ;; reduce risk of loading outdated bytecode
		(setq load-prefer-newer t))
#+end_src

** =Use-package= instructions (because I forget them)
Okay, so here's the deal. I went super gung-ho on =use-package= based
configuration for awhile. I regret that now. I regret that because it's simpler
to "learn to write emacs-lisp gud", and know how to debug that, than to use the
fancy =use-package= macro system. Now I only do fancy =use-package= stuff
by copy pasting others' nonsense, and revert to emacs lisp if it's causing probs.

_Basics For Reference_
:init - execute stuff before loading. =:init (setq my-var t)= :config - execute
stuff after loading. =:config (foo-mode t)= :bind - set keybindings. =:bind
("C-g" . keyboard-quit)= don't forget the '.' :bind-keymap - eg projectile, set
a command-map prefix.

* use sensible-defaults.el (thank you hrs)
Use [[https://github.com/hrs/sensible-defaults.el][sensible-defaults.el]] for some basic settings. This is really good to read
from if learning elisp.
** Some nice from SD:
*** functions
sensible-defaults/comment-or-uncomment-blah
sensible-defaults/reset-text-size
*** Nice settingss:
Save backups to /tmp
set default directory when opening a file to home
30x garbage collection limit
delete trailing whitespaces on saves
treat camelCaseSubWords as separate words
always follow symlinks when opening files
when saving a file starting with #!, make it executable automatically
Don't assume sentences should have two spaces after periods
When saving a file that doesn't exist, offer to create it and any super-directories necessary
Turn on transient mark mode
If text is selected and then I type text, replace the text (delete-selection-mode)
Append a newline to all saved files
Confirm whether to close emacs before closing emacs
Suppress the startup message and clear the scratch buffer
Tell dired mode to list file sizes in kb and mb etc.
Ask y/n instead of yes-or-no
Turn on syntax highlighting where possible (font-lock-mode)
refresh buffers so they don't get out of sync with auto-revert-mode
Show matching parens (show-paren-mode)
Flash screen, not audibell
Set line length to 60 *play with this number as suits me*
If clicking on stuff to open in emacs, open in existing emacs window
Insert text where the point is, not where the mouse is
*** Nice keybindings
M-; comments or uncomments a region
C-+-_= are size adjustments
** 4 lines that make all the diff
#+begin_src emacs-lisp
  (load-file "~/.emacs.d/sensible-defaults.el")
  (sensible-defaults/use-all-settings)
  (sensible-defaults/use-all-keybindings)
  (sensible-defaults/backup-to-temp-directory)
#+end_src

* Utility function friends
** All my defun friends live here
Also, some of hrs's. Good to read when learning elisp.
*** Mine
reload, open config, open zsh profile
#+begin_src emacs-lisp
	(defun reload ()
		"shorcut to reload init file"
		(interactive)
		(load-file "~/.emacs.d/init.el"))

	(defun cedit ()
		"shortcut to edit config.org file"
		(interactive)
		(find-file "~/.emacs.d/config.org"))

	(defun tedit ()
		"shortcut to edit org/tktodos.org"
		(interactive)
		(find-file "~/org/tktodos.org"))

	(defun pedit ()
		"shortcut to edit .profile"
		(interactive)
		(find-file "~/.zshrc"))

 (defun insert-date (prefix)
    "Insert the current date. With prefix-argument, use ISO format. With
   two prefix arguments, write out the day and month name."
    (interactive "P")
    (let ((format (cond
                   ((not prefix) "%Y-%m-%d")
                   ((equal prefix '(4)) "%d-%m-%Y")
                   ((equal prefix '(16)) "%A, %d. %B %Y")))
          (system-time-locale "us_US"))
      (insert (format-time-string format))))
(global-set-key (kbd "C-c d") 'insert-date)
#+end_src

*** Better than C-h <blah>, get help for emacs-lisp thing at point in tooltip. C-h C-h.
thx. Get a tooltip for thing at point.
#+begin_src emacs-lisp
(defun describe-in-popup (fn)
  (let* ((thing (symbol-at-point))
         (description (save-window-excursion
                        (funcall fn thing) ;; This is the yield point
                        (switch-to-buffer "*Help*")
                        (buffer-string))))
    (popup-tip description
               :point (point)
               :around t
               :height 30
               :scroll-bar t
               :margin t)))
(defun describe-thing-in-popup ()
  (interactive)
  (let* ((thing (symbol-at-point)))
    (cond
     ((fboundp thing) (describe-in-popup 'describe-function))
     ((boundp thing) (describe-in-popup 'describe-variable)))))

(global-set-key (kbd "C-h C-h") 'describe-thing-in-popup)
#+end_src
*** hrs's convenience functions that I mostly keep around to study
rename file, get scratch buffer, kill current buffer no prompt
#+begin_src emacs-lisp
  (defun hrs/rename-file (new-name)
    (interactive "FNew name: ")
    (let ((filename (buffer-file-name)))
      (if filename
          (progn
            (when (buffer-modified-p)
               (save-buffer))
            (rename-file filename new-name t)
            (kill-buffer (current-buffer))
            (find-file new-name)
            (message "Renamed '%s' -> '%s'" filename new-name))
        (message "Buffer '%s' isn't backed by a file!" (buffer-name)))))

  (defun hrs/generate-scratch-buffer ()
    "Create and switch to a temporary scratch buffer with a random
       name."
    (interactive)
    (switch-to-buffer (make-temp-name "scratch-")))

  (defun hrs/kill-current-buffer ()
    "Kill the current buffer without prompting."
    (interactive)
    (kill-buffer (current-buffer)))
#+end_src

Add entries to alist, sudo find
#+begin_src emacs-lisp
  (defun hrs/add-auto-mode (mode &rest patterns)
    "Add entries to `auto-mode-alist' to use `MODE' for all given file `PATTERNS'."
    (dolist (pattern patterns)
      (add-to-list 'auto-mode-alist (cons pattern mode))))

  (defun hrs/find-file-as-sudo ()
    (interactive)
    (let ((file-name (buffer-file-name)))
      (when file-name
        (find-alternate-file (concat "/sudo::" file-name)))))

  (defun hrs/region-or-word ()
    (if mark-active
        (buffer-substring-no-properties (region-beginning)
                                        (region-end))
      (thing-at-point 'word)))
#+end_src

append to path, insert generated password, display a desktop notification
#+begin_src emacs-lisp
	(defun hrs/append-to-path (path)
		"Add a path both to the $PATH variable and to Emacs' exec-path."
		(setenv "PATH" (concat (getenv "PATH") ":" path))
		(add-to-list 'exec-path path))

	(defun hrs/insert-password ()
		(interactive)
		(shell-command "pwgen 30 -1" t))

	(defun hrs/notify-send (title message)
		"Display a desktop notification by shelling out to `notify-send'."
		(call-process-shell-command
		 (format "notify-send -t 2000 \"%s\" \"%s\"" title message)))
#+end_src

** Macro game going strong
My lisp and rust block macros
#+begin_src emacs-lisp
(fset 'tk-org-insert-lisp-block
   "#+begin_src emacs-lisp\C-m\C-m#+end_src\C-p")
(global-set-key (kbd "<f2>") 'tk-org-insert-lisp-block)
#+end_src

#+begin_src emacs-lisp
(fset 'tk-org-insert-rust-block
   "#+begin_src rust\C-m\C-m#+end_src\C-p")
(global-set-key (kbd "<f3>") 'tk-org-insert-rust-block)
#+end_src

** Handy kbd shortcuts live here
Return and indent instead of just return
#+begin_src emacs-lisp
(define-key global-map (kbd "RET") 'newline-and-indent)
(define-key global-map (kbd "<f7>") 'eshell)
#+end_src

* Prettify emacs window appearance and general settings
** No Backups~
#+begin_src emacs-lisp
	(setq make-backup-files nil) ; none of these~
	(setq auto-save-default t)
#+end_src
** Theme
#+begin_src emacs-lisp
	(use-package solarized-theme
	 :config (load-theme 'solarized-gruvbox-dark t))
	;; make src block code look like normal text
	(add-hook 'text-mode-hook
						 (lambda ()
							(variable-pitch-mode 1)))
#+end_src
** Window configuration
*** Zoom mode
[[https://github.com/cyrus-and/zoom][Zoom]]. Auto resize windows on active buffer switch.
#+begin_src emacs-lisp
	(use-package zoom
		:config (zoom-mode t))
#+end_src
*** =subword-mode= to tell emacs to treat CamelCaseAsDifferentWords on M-f/b
#+begin_src emacs-lisp
	(setq subword-mode t)
#+end_src
*** Autofill, electric pair, linum, highlighting
line numberings. And highlighting current line. and highlighting differences on
git tracked files.
[[https://github.com/dgutov/diff-hl][diff-hl-mode]]
#+begin_src emacs-lisp
	(add-hook 'text-mode-hook 'turn-on-auto-fill)
	(global-linum-mode 1)
	(global-hl-line-mode)
	(setq electric-pair-mode 1)
	(use-package diff-hl)
	(global-diff-hl-mode)
#+end_src

*** Remove things
Get rid of the annoying parts of the display.
#+begin_src emacs-lisp
  (tool-bar-mode 0)
  (menu-bar-mode 0)
  (scroll-bar-mode 0)
  (set-window-scroll-bars (minibuffer-window) nil nil) ; minibuffer window has a scroll bar for some reason
#+end_src

** Use fancy lambdas, =moody= and =minions=
lambdas: prettify symbols, column and line number
#+begin_src emacs-lisp
  (global-prettify-symbols-mode t)
  (setq column-number-mode t)
  (setq line-number-mode t)
#+end_src

[[https://github.com/tarsius/moody][Moody]] is a pretty mode bar
[[https://github.com/tarsius/minions][minions]] mode changes what minor are listed in the bar
#+begin_src emacs-lisp
	(use-package moody
		:config
		(setq x-underline-at-descent-line t)
		(moody-replace-mode-line-buffer-identification)
		(moody-replace-vc-mode))
	(use-package minions)
	(minions-mode 1)
#+end_src

** [off] Turn on flyspell check in text mode
[[https://www.emacswiki.org/emacs/FlySpell][emacswiki flyspell]]
;#+begin_src emacs-lisp
	(dolist (hook '(text-mode-hook)) ; when entering text mode
				(add-hook hook (lambda () (flyspell-mode 1)))) ; add hook to turn on flyspell
	;(add-hook 'prog-mode-hook ; turn on flyspell in comments of programming modes
	;					(lambda ()
	;						(flyspell-prog-mode)
	;					))
	(setq flyspell-issue-message-flag nil) ; printing messages for every word slows down perf
;#+end_src
Toy with this for awhile, then try out [[https://github.com/d12frosted/flyspell-correct][flyspell-correct]] and [[https://github.com/d12frosted/flyspell-correct#flyspell-correct-ivy-interface][ivy interface]] later
[[http://ergoemacs.org/emacs/emacs_spell_checker_problems.html][Xah has problems and solutions with flyspell if these don't work.]]
** work groups for remembering window setup ; Cc z [c/v/Cs/Cf]
Want emacs to remember my window setup for stuff. Storing views in registers
aren't persistent across sessions. [[https://github.com/pashinin/workgroups2][docs]].
Note that this can get buggy with treemacs, beware.
; #+begin_src emacs-lisp
	(use-package workgroups2
		:config
	(setq wg-session-file "~/.emacs.d/workgroups")
	(setq wg-emacs-exit-save-behavior 'save))      ; Options: 'save 'ask nil
	(setq wg-prefix-key (kbd "C-c z"))
	(workgroups-mode)
;#+end_src

* Cool Tools
** =ivy-mode=, with =counsel= and =swiper= - navigation and other things
Ivy and helm do similar stuff, move ya from place to place and complete stuff.
Ivy claims to be more minimal.
Using [[https://oremacs.com/swiper/][oremacs]] user manual heavily.
[[https://writequit.org/denver-emacs/presentations/2017-04-11-ivy.html][this link's much nicer tho]]
New learn: use C-c v/V to store the current set of windows open.
*** Standard commands mostly replace standards. Ch i/l info-lookup/library
Copy pasted from docs.
#+begin_src emacs-lisp
(use-package ivy)
(use-package swiper) ; search extension to ivy
(use-package counsel) ;
(ivy-mode 1) ; globally turn on ivy
(setq ivy-use-virtual-buffers t) ; variably sized
(setq ivy-count-format "(%d/%d) ")
(global-set-key (kbd "C-s") 'swiper-isearch)
(global-set-key (kbd "C-r") 'swiper-isearch-backward)
(global-set-key (kbd "M-x") 'counsel-M-x)
(global-set-key (kbd "C-x C-f") 'counsel-find-file)
(global-set-key (kbd "M-y") 'counsel-yank-pop) ; nicer kill ring
(global-set-key (kbd "C-h f") 'counsel-describe-function)
(global-set-key (kbd "C-h v") 'counsel-describe-variable)
(global-set-key (kbd "C-h l") 'counsel-find-library)
(global-set-key (kbd "C-h i") 'counsel-info-lookup-symbol)
(global-set-key (kbd "C-x b") 'ivy-switch-buffer)
(global-set-key (kbd "C-c v") 'ivy-push-view)
(global-set-key (kbd "C-c V") 'ivy-pop-view)
#+end_src

Ivy views store the set of buffers open in the current frame.
*** Shell and System tools - C-c k rg, C-c j file jump
#+begin_src emacs-lisp
(global-set-key (kbd "C-c k") 'counsel-rg)
(global-set-key (kbd "C-c j") 'counsel-file-jump)
#+end_src

*** Ivy-resume (Cc r), bkmk (Cc b), what the hell binding is this (Cc d), jump to a heading Cc o
#+begin_src emacs-lisp
(global-set-key (kbd "C-c r") 'ivy-resume)
;(global-set-key (kbd "C-c b") 'counsel-bookmark) ; weird stuff goin on
(global-set-key (kbd "C-c o") 'counsel-outline)
#+end_src

*** Ivy-rich
[[https://github.com/Yevgnen/ivy-rich][ivy rich]], in combination with ivy descbind (Cc d) and which-key, installed
earlier, make it wayyyy easier to discover and remember what my keybinds are.
#+begin_src emacs-lisp
(use-package ivy-rich
 :config (ivy-rich-mode 1))
(setcdr (assq t ivy-format-functions-alist) #'ivy-format-function-line) ; recommended format
(setq ivy-rich-path-style 'abbrev) ; abbreviate paths with ~/
#+end_src

*** =Avy= M-t
jump to visible text.
#+begin_src emacs-lisp
(use-package avy)
(global-set-key (kbd "M-t") 'avy-goto-word-1)
#+end_src

** =Counsel-projectile= - jump around project files, Ccp m, Ccp p
[[https://github.com/bbatsov/projectile][Projectile]]: finding and moving around .git or .projectile controlled project
files. Projectile-commander is useful magit-like.
#+begin_src emacs-lisp
(use-package projectile)
(use-package counsel-projectile)
(counsel-projectile-mode)
(define-key projectile-mode-map (kbd "C-c p") 'projectile-command-map)
#+end_src
** =magit= - git version control inside emacs. C-x g
[[https://github.com/magit/magit][magit]]. Interface to git. [[https://magit.vc/manual/forge/][Forge]] to interface with github. Sorta shitty
documentation on forge. Trying [[https://gist.github.com/bmhatfield/cc21ec0a3a2df963bffa3c1f884b676b][this]] solution. Hoo baby it worked.

config.el
(require 'use-package-ensure)
(setq use-package-always-ensure t)

(use-package auto-compile
	:config	(auto-compile-on-save-mode) ;; reduce risk of loading outdated bytecode
	(setq load-prefer-newer t))

(load-file "~/.emacs.d/sensible-defaults.el")
(sensible-defaults/use-all-settings)
(sensible-defaults/use-all-keybindings)
(sensible-defaults/backup-to-temp-directory)

(get-buffer-create buffer)
(with-current-buffer)
(or nil nil nil (message "hi") (message "bye"))


(defun reload ()
		"shorcut to reload init file"
		(interactive)
		(load-file "~/.emacs.d/init.el"))

	(defun cedit ()
		"shortcut to edit config.org file"
		(interactive)
		(find-file "~/.emacs.d/config.org"))

	(defun tedit ()
		"shortcut to edit org/tktodos.org"
		(interactive)
		(find-file "~/org/tktodos.org"))

	(defun pedit ()
		"shortcut to edit .profile"
		(interactive)
		(find-file "~/.zshrc"))

 (defun insert-date (prefix)
    "Insert the current date. With prefix-argument, use ISO format. With
   two prefix arguments, write out the day and month name."
    (interactive "P")
    (let ((format (cond
                   ((not prefix) "%Y-%m-%d")
                   ((equal prefix '(4)) "%d-%m-%Y")
                   ((equal prefix '(16)) "%A, %d. %B %Y")))
          (system-time-locale "us_US"))
      (insert (format-time-string format))))
(global-set-key (kbd "C-c d") 'insert-date)

(defun describe-in-popup (fn)
  (let* ((thing (symbol-at-point))
         (description (save-window-excursion
                        (funcall fn thing) ;; This is the yield point
                        (switch-to-buffer "*Help*")
                        (buffer-string))))
    (popup-tip description
               :point (point)
               :around t
               :height 30
               :scroll-bar t
               :margin t)))
(defun describe-thing-in-popup ()
  (interactive)
  (let* ((thing (symbol-at-point)))
    (cond
     ((fboundp thing) (describe-in-popup 'describe-function))
     ((boundp thing) (describe-in-popup 'describe-variable)))))

(global-set-key (kbd "C-h C-h") 'describe-thing-in-popup)

(defun hrs/rename-file (new-name)
  (interactive "FNew name: ")
  (let ((filename (buffer-file-name)))
    (if filename
        (progn
          (when (buffer-modified-p)
             (save-buffer))
          (rename-file filename new-name t)
          (kill-buffer (current-buffer))
          (find-file new-name)
          (message "Renamed '%s' -> '%s'" filename new-name))
      (message "Buffer '%s' isn't backed by a file!" (buffer-name)))))

(defun hrs/generate-scratch-buffer ()
  "Create and switch to a temporary scratch buffer with a random
     name."
  (interactive)
  (switch-to-buffer (make-temp-name "scratch-")))

(defun hrs/kill-current-buffer ()
  "Kill the current buffer without prompting."
  (interactive)
  (kill-buffer (current-buffer)))

(defun hrs/add-auto-mode (mode &rest patterns)
  "Add entries to `auto-mode-alist' to use `MODE' for all given file `PATTERNS'."
  (dolist (pattern patterns)
    (add-to-list 'auto-mode-alist (cons pattern mode))))

(defun hrs/find-file-as-sudo ()
  (interactive)
  (let ((file-name (buffer-file-name)))
    (when file-name
      (find-alternate-file (concat "/sudo::" file-name)))))

(defun hrs/region-or-word ()
  (if mark-active
      (buffer-substring-no-properties (region-beginning)
                                      (region-end))
    (thing-at-point 'word)))

(defun hrs/append-to-path (path)
	"Add a path both to the $PATH variable and to Emacs' exec-path."
	(setenv "PATH" (concat (getenv "PATH") ":" path))
	(add-to-list 'exec-path path))

(defun hrs/insert-password ()
	(interactive)
	(shell-command "pwgen 30 -1" t))

(defun hrs/notify-send (title message)
	"Display a desktop notification by shelling out to `notify-send'."
	(call-process-shell-command
	 (format "notify-send -t 2000 \"%s\" \"%s\"" title message)))

(fset 'tk-org-insert-lisp-block
   "#+begin_src emacs-lisp\C-m\C-m#+end_src\C-p")
(global-set-key (kbd "<f2>") 'tk-org-insert-lisp-block)

(fset 'tk-org-insert-rust-block
   "#+begin_src rust\C-m\C-m#+end_src\C-p")
(global-set-key (kbd "<f3>") 'tk-org-insert-rust-block)

(define-key global-map (kbd "RET") 'newline-and-indent)
(define-key global-map (kbd "<f7>") 'eshell)

(setq make-backup-files nil) ; none of these~
(setq auto-save-default t)

(use-package solarized-theme
 :config (load-theme 'solarized-gruvbox-dark t))
;; make src block code look like normal text
(add-hook 'text-mode-hook
					 (lambda ()
						(variable-pitch-mode 1)))

(use-package zoom
	:config (zoom-mode t))

(setq subword-mode t)

(add-hook 'text-mode-hook 'turn-on-auto-fill)
(global-linum-mode 1)
(global-hl-line-mode)
(setq electric-pair-mode 1)
(use-package diff-hl)
(global-diff-hl-mode)

(tool-bar-mode 0)
(menu-bar-mode 0)
(scroll-bar-mode 0)
(set-window-scroll-bars (minibuffer-window) nil nil) ; minibuffer window has a scroll bar for some reason

(global-prettify-symbols-mode t)
(setq column-number-mode t)
(setq line-number-mode t)

(use-package moody
	:config
	(setq x-underline-at-descent-line t)
	(moody-replace-mode-line-buffer-identification)
	(moody-replace-vc-mode))
(use-package minions)
(minions-mode 1)

(use-package ivy)
(use-package swiper) ; search extension to ivy
(use-package counsel) ;
(ivy-mode 1) ; globally turn on ivy
(setq ivy-use-virtual-buffers t) ; variably sized
(setq ivy-count-format "(%d/%d) ")
(global-set-key (kbd "C-s") 'swiper-isearch)
(global-set-key (kbd "C-r") 'swiper-isearch-backward)
(global-set-key (kbd "M-x") 'counsel-M-x)
(global-set-key (kbd "C-x C-f") 'counsel-find-file)
(global-set-key (kbd "M-y") 'counsel-yank-pop) ; nicer kill ring
(global-set-key (kbd "C-h f") 'counsel-describe-function)
(global-set-key (kbd "C-h v") 'counsel-describe-variable)
(global-set-key (kbd "C-h l") 'counsel-find-library)
(global-set-key (kbd "C-h i") 'counsel-info-lookup-symbol)
(global-set-key (kbd "C-x b") 'ivy-switch-buffer)
(global-set-key (kbd "C-c v") 'ivy-push-view)
(global-set-key (kbd "C-c V") 'ivy-pop-view)

(global-set-key (kbd "C-c k") 'counsel-rg)
(global-set-key (kbd "C-c j") 'counsel-file-jump)

(global-set-key (kbd "C-c r") 'ivy-resume)
;(global-set-key (kbd "C-c b") 'counsel-bookmark) ; weird stuff goin on
(global-set-key (kbd "C-c o") 'counsel-outline)

(use-package ivy-rich
 :config (ivy-rich-mode 1))
(setcdr (assq t ivy-format-functions-alist) #'ivy-format-function-line) ; recommended format
(setq ivy-rich-path-style 'abbrev) ; abbreviate paths with ~/

(use-package avy)
(global-set-key (kbd "M-t") 'avy-goto-word-1)

(use-package projectile)
(use-package counsel-projectile)
(counsel-projectile-mode)
(define-key projectile-mode-map (kbd "C-c p") 'projectile-command-map)

(use-package magit)
(use-package forge)

(use-package rg)
(hrs/append-to-path "/usr/local/bin") ; oddly wasn't globally in path, fixing that

(add-hook 'dired-load-hook
          (lambda ()
            (load "dired-x")))

(use-package dumb-jump)
(global-set-key (kbd "C-M-s") 'dumb-jump-go)
(global-set-key (kbd "C-M-r") 'dumb-jump-back)
(setq dumb-jump-force-searcher 'rg)
(setq dumb-jump-selector 'ivy)

(use-package re-builder)
(setq reb-re-syntax 'string)

(use-package simpleclip)
(simpleclip-mode 1)

(use-package which-key)

(global-set-key (kbd "C-c l") 'org-store-link)
(global-set-key (kbd "C-c a") 'org-agenda)
(global-set-key (kbd "C-c c") 'org-capture)

(setq org-src-fontify-natively t)
(setq org-src-tab-acts-natively t)
(setq org-adapt-indentation nil)
;(setq org-pretty-entities nil) ; quick latex-ify in org files; annoying in codesnippets

(setq org-directory "~/org")
	(setq org-todo-keywords									; ! = timestamp, @ = create note
				'((sequence "TODO(t!)" "NOW(n!)" "WAITING(w)" "|" "DONE(d)" "CANCELED(c)")))
(setq org-log-done 'time) ; log when finished

(use-package org-download)

(setq org-default-notes-file (concat org-directory "~/org/tktodos.org")) ; capture
	(setq org-capture-templates
				 '(
		 ("z" "Misc todo" entry (file+headline "~/org/misc.org" "Misc")
			"* TODO \t %? :MISC:\nAdded: %u:" :empty-lines 1 )
		 ("d" "Dev" entry (file+headline "~/org/dev.org" "Dev")
			"* TODO \t %? :DEV:\nAdded: %u" :empty-lines 1 )
		 ("M" "Main Dev" entry (file+headline "~/org/main.org" "Main")
			"* TODO [#A] \t %? :MAIN:DEV:\nAdded: %u" :empty-lines 1 )
		 ("R" "Main Rsch" entry (file+headline "~/org/main.org" "Main")
			"* TODO [#A] \t %? :MAIN:RSCH:\nAdded: %u" :empty-lines 1 )
		 ("e" "Emacs" entry (file+headline "~/org/emacs.org" "Emacs")
			"* TODO \t %? :EMACS:\nAdded: %u" :empty-lines 1 )
		 ("p" "Personal" entry (file+headline "~/org/pers.org" "Pers")
			"* TODO \t %? :PERS:\nAdded: %u" :empty-lines 1 )
		 ("r" "Research" entry (file+headline "~/org/rsch.org" "Rsch")
			"* TODO \t %? :RSCH:\nAdded: %u" :empty-lines 1 )
		 ("i" "Idea" entry (file "~/org/ideas.org")
			"* \t %? :IDEA:\nAdded: %u" )
))

(add-hook 'org-mode-hook
			(lambda ()
				(local-set-key (kbd "C-c C-x C-l") 'org-clock-in-last)
))

(setq org-clock-idle-time 15) ;prompt after 15 idle minutes.

(setq org-agenda-files '("~/org" ))

; tf not used heavily atm
(setq org-agenda-custom-commands ; options - todo, tags, tags-todo
			'(("d" "Dev" tags-todo "DEV")
				("e" "Emacs" tags-todo "EMACS")
				("p" "Personal" tags-todo "PERS")
				("r" "Research" tags-todo "RSCH")
				("m" "Research" tags-todo "MAIN")
				))
(setq org-agenda-start-on-weekday nil) ; start today

(setq org-tag-alist '(("dev" . d) ("personal" . ?p) ("research" . ?r) ("main" . ?m)))

(use-package org-roam
					:hook
					(after-init . org-roam-mode)
					:custom ; adjust graph dot executable
					(org-roam-directory "~/org/roam")
					:bind (:map org-roam-mode-map (
									 ("C-c n f" . org-roam-find-file)
									 ("C-c n g" . org-roam-graph))
									:map org-mode-map
									(("C-c n i" . org-roam-insert))
									(("C-c n c" . org-roam-capture))
								(("C-c n I" . org-roam-insert-immediate))
								))
(use-package company-org-roam)
(setq org-roam-completion-system 'ivy)

(setq org-roam-capture--file-name-default "<%Y-%m%-%d>")
(setq org-roam-capture-templates
			 ;; '(("p" "paper" plain (function org-roam--capture-get-point)
			 ;; 	 "%?"
			 ;; 	 :file-name "paper/${topic}/${subtopic}/${slug}"
			 ;; 	 :head: "#+title: ${title}\n"
			 ;; 	 :unnarrowed t)
			 '(("w" "web" plain (function org-roam--capture-get-point)
				"%?"
				:file-name "web/${topic}/${subtopic}/${slug}"
				:head "#+title: ${title}\n"
				:unnarrowed t
				)
				 ))

(setq org-roam-graph-executable "/usr/local/bin/dot")
(use-package graphviz-dot-mode
  :config
(setq graphviz-dot-indent-width 4))
;(setq org-roam-graph-viewer "/Applications/Safari.app/Contents/MacOS/safari")

(setq-default tab-width 2)

(use-package flycheck
  :ensure t
  :init (global-flycheck-mode)) ; test

(setq lsp-keymap-prefix "M-n")
	(use-package lsp-mode
			:hook (rustic-mode . lsp)
;			:hook (sh-mode . lsp)
			:hook(go-mode . lsp)
			:commands lsp)
	;; optionally
	(use-package lsp-ui :commands lsp-ui-mode)
	(use-package lsp-ivy :commands lsp-ivy-workspa ce-symbol)
	(use-package lsp-treemacs :commands lsp-treemacs-errors-list)
	(use-package dap-mode) ; debugger - no dap-rust yet
	;; (use-package dap-LANGUAGE) to load the dap adapter for your language

(use-package rustic)  ; many nifty convenience functions. defaults to rust-analyzer > rls
(setq rustic-lsp-server 'rust-analyzer)
(setq lsp-rust-analyzer-server-command '("~/.cargo/bin/rust-analyzer"))
(custom-set-faces
  '(rustic-compilation-error ((t (:foreground "Red"))))
  '(rustic-compilation-warning ((t (:foreground "Red"))))
  '(rustic-compilation-message ((t (:foreground "Red"))))
  '(rustic-compilation-info ((t (:foreground "Blue"))))
  '(rustic-compilation-line ((t (:foreground "Blue")))))

(use-package rust-mode
  :config
	      (hrs/append-to-path "~/.cargo/bin")
  (setq rust-format-on-save t))

(use-package cargo)
(use-package toml-mode)

; testing, possibly useful on Starchy
      ;(setenv "RUST_SRC_PATH" "/home/thor/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library")
      (use-package racer
      :config (setq company-tooltip-align-annotations t)
      :hook ((rust-mode . racer-mode)
      (rust-mode . rustic-mode)
      (add-racer-mode . eldoc-mode) ; shows in echo area the arg list of the fn at point
      (racer-mode . company-mode)) ; company autocomplete sometimes slows editor down significantly
      :bind (:map rust-mode-map ("TAB" . company-indent-or-complete-common)))

(use-package flycheck-rust) ; runs on save buffer
(with-eval-after-load 'rust-mode
		(add-hook 'flycheck-mode-hook 'flycheck-rust-setup))
(add-hook 'rust-mode-hook
		(lambda ()
		(setq cargo-minor-mode t) ; Cc Cc C(b/r/t)
	))

(use-package rust-playground)

(use-package ob-rust)

(use-package go-mode)
(use-package go-errcheck)
(use-package company) ; autocompletes
(use-package company-go)
(setq company-tooltip-limit 20)     ; bigger popup window
(setq company-idle-delay .2)        ; decrease delay before autocompletion popup shows

(setenv "GOPATH" "/Users/thor/go")
(hrs/append-to-path (concat (getenv "GOPATH") "/bin")) ; user gopath
(hrs/append-to-path "/usr/local/go/bin") ; other shit that we like

(setq gofmt-command "goimports")
(add-hook 'before-save-hook 'gofmt-before-save)

(add-hook 'go-mode-hook
						(lambda ()
							(if (not (string-match "go" compile-command))
									(set (make-local-variable 'compile-command)
										 "go run ")
								)))

(add-hook 'go-mode-hook
					(lambda ()
						(set (make-local-variable 'company-backends)
								 '(company-go))
						(company-mode)
						(flycheck-mode)
						(local-set-key (kbd "C-c C-c C-r") 'compile)
	))

(use-package paredit)
(use-package rainbow-delimiters)

(setq lispy-mode-hooks
      '(emacs-lisp-mode-hook
        lisp-mode-hook
        scheme-mode-hook))
(dolist (hook lispy-mode-hooks)
  (add-hook hook (lambda ()
                   (setq show-paren-style 'expression)
                   (paredit-mode)
                   (rainbow-delimiters-mode))))

(use-package eldoc
  :config
  (add-hook 'emacs-lisp-mode-hook 'eldoc-mode))

(use-package flycheck-package) ; should b called emacs-lisp flycheck
(eval-after-load 'flycheck
  '(flycheck-package-setup))

(use-package solidity-mode)
; optionals
(setq solidity-comment-style 'slash) ; 'star by default
; optional keybinds
(define-key solidity-mode-map (kbd "C-c C-g") 'solidity-estimate-gas-at-point)
; provide path to solc and solium binaries ;  these are overlapping checkers and can be used simultaneously
; (setq solidity-solc-path "/bin")
; (setq solidity-solium-path "/home/lefteris/.npm-global/bin/solium")
;(use-package 'solidity-flycheck)
;(setq solidity-flycheck-solc-checker-active t)
;(setq solidity-flycheck-solium-checker-active t)
; (setq flycheck-solidity-solc-addstd-contracts t) ; enable to include standard contracts
(use-package company-solidity)
(add-hook 'solidity-mode-hook
	(lambda ()
	(set (make-local-variable 'company-backends)
		(append '((company-solidity company-capf company-dabbrev-code))
			company-backends))))

(use-package company-shell)
(add-to-list 'company-backends 'company-shell)

;  (use-package auctex)
; Error (use-package): ; auctex/:catch: Loading file /Users/thor/.emacs.d/elpa/auctex-12.2.4/auctex.elc failed to provide feature auctex
	(setq TeX-auto-save t)
	(setq TeX-parse-self t)
	(setq-default TeX-master nil)
	(add-hook 'LaTeX-mode-hook 'visual-line-mode) ; an altern to auto-fill-mode
	(add-hook 'LaTeX-mode-hook 'flyspell-mode)
	(add-hook 'LaTeX-mode-hook 'LaTeX-math-mode)
	(add-hook 'LaTeX-mode-hook 'turn-on-reftex)
	(setq reftex-plug-into-AUCTeX t)
	(add-hook 'tex-mode-hook
						(lambda ()
						(latex-electric-env-pair-mode)))

(defun set-frame-size-according-to-resolution ()
  (interactive)
  (if window-system
  (progn
    ;; use 120 char wide window for largeish displays
    ;; and smaller 80 column windows for smaller displays
    ;; pick whatever numbers make sense for you
    (if (> (x-display-pixel-width) 1280)
           (add-to-list 'default-frame-alist (cons 'width 100))
           (add-to-list 'default-frame-alist (cons 'width 80)))
    ;; for the height, subtract a couple hundred pixels
    ;; from the screen height (for panels, menubars and
    ;; whatnot), then divide by the height of a char to
    ;; get the height we want
    (add-to-list 'default-frame-alist
         (cons 'height (/ (- (x-display-pixel-height) 60) ; close as I can get to full left half
                             (frame-char-height)))))))

(set-frame-size-according-to-resolution)
